torch/_inductor/ir.py

```diff
@@ -6684,6 +6684,220 @@ def codegen(self, wrapper):
+            combine_subgraph=combine_subgraph,
+            init=init,
+            xs=xs,
+            dim=dim,
+            reverse=reverse,
+            additional_inputs=additional_inputs,
+            layout=MultiOutputLayout(device),
+            iter_idx_expr=iter_idx_expr,
+        )
+
+        # last_carry should have the same shape and stride as init
+        last_carry_output = [
+            MultiOutput(
+                FixedLayout(
+                    device=ini.get_device(),
+                    dtype=ini.get_dtype(),
+                    size=ini.get_size(),
+                    stride=ini.get_stride(),
+                ),
+                sequential_scan,
+                [(list, i)],
+            )
+            for i, ini in enumerate(init)
+        ]
+
+        def stacked_size_stride(y, scan_length):
+            sizes = [scan_length, *y.get_size()]
+            # This is a hacky way to get the stride of the stacked output
+            # Semantic-wise this is correct because scan will torch.stack the intermediates, so
+            # torch.stack will force the output become contingous.
+            return sizes, FlexibleLayout.contiguous_strides(sizes)
+
+        stacked_y_size_stride = [stacked_size_stride(y, scan_length) for y in ys]
+
+        stacked_output = [
+            MultiOutput(
+                FixedLayout(
+                    device=y.get_device(),
+                    dtype=y.get_dtype(),
+                    size=stacked_y_size_stride[i][0],
+                    stride=stacked_y_size_stride[i][1],
+                ),
+                sequential_scan,
+                [(list, i + num_init_leaves)],
+            )
+            for i, y in enumerate(ys)
+        ]
+
+        outputs = [*last_carry_output, *stacked_output]
+        sequential_scan.outputs = outputs
+        return outputs
+
+    def codegen(self, wrapper):
+        wrapper.codegen_sequential_scan(self)
+
+    # The signature before the transformation:
+    # gm(*init: List[Tensor],
+    #   *selected_inputs: List[Tensor],
+    #   *additional_inputs: List[Tensor]
+    # ) -> last_carry: List[Tensor], ys: List[Tensor]
+    #
+    # The signature after _to_out_variant_graph:
+    # gm_out_variant(*init: List[Tensor],
+    #   *xs: List[Tensor],
+    #   *additional_inputs: List[Tensor],
+    #   *ys_outs: List[Tensor]
+    #   *idx: SymInt,
+    # ) -> last_carry: List[Tensor], ys: List[Tensor]
+    #
+    # For example, if scan_length is 10, selected_inputs[0].shape is (3, 4), ys[0].shape is (5, 6)
+    # In gm_out_variant's input, xs[0].shape is (10, 3, 4), ys_outs[0].shape is (10, 5, 6).
+    # idx is a symint that's computed based on the iteration idx of the loop in wrapper codegen.
+    #
+    # Note: even though ys are copied into ys_outs directly and is not going to be used downsteram,
+    # we didn't delete them in the out variant graph in case they gets DCEed accidentally.
+    @staticmethod
+    def _to_out_variant_graph(
```

Comment:
I'm not sure this is the best place for this transformation.. Also, would it be easier to maintain this specified as [replace_by_example](https://github.com/pytorch/pytorch/blob/08dba25775afc2d979ecab9eb9a052ebf4483bd3/torch/_inductor/comms.py#L636) instead of manually constructing the graph and manually inserting metas with clone (which may not replicate aliasing, for instance, or the correct striding)

Reply:
yeah, that sounds a better idea. Let me check how to do this properly with `replace_by_example`. Thanks for the code pointer!